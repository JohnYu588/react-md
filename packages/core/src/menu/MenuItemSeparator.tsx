import type { HTMLAttributes } from "react";
import { forwardRef } from "react";
import type { DividerProps, VerticalDividerProps } from "../divider";
import { divider, useVerticalDividerHeight } from "../divider";
import { useMenuConfiguration } from "./MenuConfigurationProvider";

/** @remarks \@since 5.0.0 */
export interface MenuItemSeparatorProps
  extends HTMLAttributes<HTMLLIElement>,
    Pick<DividerProps, "inset" | "vertical">,
    Pick<VerticalDividerProps, "maxHeight"> {}

/**
 * This component renders a `<li role="separator">` with the divider styles. It
 * will also automatically render itself vertically instead of horizontally if
 * the menu is rendering horizontally.
 *
 * @remarks \@since 5.0.0 Renders as an `<li>` instead of a `<div>` or `<hr />`.
 */
export const MenuItemSeparator = forwardRef<
  HTMLLIElement,
  MenuItemSeparatorProps
>(function MenuItemSeparator(props, ref) {
  const {
    style,
    className,
    maxHeight,
    inset,
    vertical: propVertical,
    children,
    ...remaining
  } = props;

  const horizontal = useMenuConfiguration().horizontal;
  const vertical = propVertical ?? horizontal;
  const heightProps = useVerticalDividerHeight({
    ref,
    style,
    maxHeight: vertical ? maxHeight ?? 1 : 0,
  });

  return (
    <li
      {...remaining}
      {...heightProps}
      aria-orientation={vertical ? "vertical" : undefined}
      role="separator"
      className={divider({ inset: inset && !vertical, vertical, className })}
    >
      {children}
    </li>
  );
});
