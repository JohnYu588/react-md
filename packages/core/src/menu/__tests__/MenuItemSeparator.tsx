import { describe, expect, it } from "@jest/globals";
import { render, screen } from "@testing-library/react";
import { createRef } from "react";

import { MenuItemSeparator } from "../MenuItemSeparator.js";

describe("MenuItemSeparator", () => {
  it("should apply the correct styling, HTMLAttributes, and allow a ref", () => {
    const ref = createRef<HTMLLIElement>();
    const props = {
      ref,
    } as const;

    const { rerender } = render(<MenuItemSeparator {...props} />);

    const element = screen.getByRole("separator");
    expect(ref.current).toBeInstanceOf(HTMLLIElement);
    expect(ref.current).toBe(element);
    expect(element).toMatchSnapshot();

    rerender(<MenuItemSeparator inset {...props} />);
    expect(element).toMatchSnapshot();

    rerender(<MenuItemSeparator inset vertical {...props} />);
    expect(element).toMatchSnapshot();

    rerender(<MenuItemSeparator vertical {...props} />);
    expect(element).toMatchSnapshot();

    rerender(
      <MenuItemSeparator
        {...props}
        style={{ color: "white" }}
        className="custom-class-name"
      />
    );
    expect(element).toMatchSnapshot();
  });
});
