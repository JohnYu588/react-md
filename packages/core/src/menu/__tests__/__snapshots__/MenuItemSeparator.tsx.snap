// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MenuItemSeparator should apply the correct styling, HTMLAttributes, and allow a ref 1`] = `
<li
  class="rmd-divider"
  role="separator"
/>
`;

exports[`MenuItemSeparator should apply the correct styling, HTMLAttributes, and allow a ref 2`] = `
<li
  class="rmd-divider rmd-divider--inset"
  role="separator"
/>
`;

exports[`MenuItemSeparator should apply the correct styling, HTMLAttributes, and allow a ref 3`] = `
<li
  aria-orientation="vertical"
  class="rmd-divider rmd-divider--vertical"
  role="separator"
/>
`;

exports[`MenuItemSeparator should apply the correct styling, HTMLAttributes, and allow a ref 4`] = `
<li
  aria-orientation="vertical"
  class="rmd-divider rmd-divider--vertical"
  role="separator"
/>
`;

exports[`MenuItemSeparator should apply the correct styling, HTMLAttributes, and allow a ref 5`] = `
<li
  class="rmd-divider custom-class-name"
  role="separator"
  style="color: white;"
/>
`;
